djangocms_form_builder/cms_plugins/ajax_plugins.py

```diff
@@ -240,11 +240,17 @@ def render(self, context, instance, placeholder):
         form = self.get_ajax_form()
         context.update(self.set_context(context, instance, placeholder))
         context["form_counter"] = context.get("form_counter", 0) + 1
-        context.update({
-            "instance": instance,
-            "form": form,
-            "uid": f"{instance.id}{getattr(form, 'slug', '')}-{context['form_counter']}",
-        })
+        has_submit_button = any(
+            child.plugin_type == "SubmitPlugin" for child in instance.get_children()
```

Comment:
@earthcomfy This works but probably is inefficient, since it hits the database. When rendering plugins, django CMS first takes all plugins of a placeholder and "downcasts" all of them. Downcasting is the process of taking the plugin tree (`CMSPlugin` instances) and turning them into their respecitve real classes, such as `Text`, or `Form`. 

This also means that all children have been read from the db already when `render` is called. They recursively are part of the list ``self.child_plugin_instances` . Can you evaluate this instead?

Reply:
I see. I have updated it to use that.

Reply:
That's going into the right direction. The thing is that `child_plugin_instances` only includes **direct** children. So you're missing out on indirect submit buttons. The children also have `child_plugin_instances`. Probably it makes sense to put the test into a separate method.

Reply:
hmm, I think the code already uses `child_plugin_instances` that were already recursively loaded by `traverse`. I tried adding a test to reproduce this but it doesn't fail.